### ISSUE TITLE
Is this a safe way to use ulid concurrently with other libraries too?

### ISSUE THREAD

**Issue by frederikhors** [oklog/ulid#79]
_I just asked [this question on SO](https://stackoverflow.com/questions/72584598/is-this-a-safe-way-to-use-ulid-concurrently-with-other-libraries-too). I'm posting this here too both for me and for for future "researchers"._

I'm trying to use for the first time the [ulid package](https://github.com/oklog/ulid).

In [their README](https://github.com/oklog/ulid#usage) they say:

> Please note that `rand.Rand` from the `math` package is *not* safe for concurrent use.
> Instantiate one per long living go-routine or use a `sync.Pool` if you want to avoid the potential contention of a locked `rand.Source` as its been frequently observed in the package level functions.

Can you help me understand what does this mean and **how to write SAFE code for concurrent use with libraries such [ent](https://github.com/ent/ent) or [gqlgen](https://github.com/99designs/gqlgen)**?

Example: I'm using the below code in my app to generate new IDs (sometimes even many of them in the same millisecond which is fine for `ulid`).

```go
import (
  "math/rand"
  "time"

  "github.com/oklog/ulid/v2"
)

var defaultEntropySource *ulid.MonotonicEntropy

func init() {
  defaultEntropySource = ulid.Monotonic(rand.New(rand.NewSource(time.Now().UnixNano())), 0)
}

func NewID() string {
  return ulid.MustNew(ulid.Timestamp(time.Now()), defaultEntropySource).String()
}
```

**Is this a safe way to use the package?**

Can you help me better understand?

**Comment by tsenart** [comment]
> Can you help me understand what does this mean and how to write SAFE code for concurrent use with libraries such [ent](https://github.com/ent/ent) or [gqlgen](https://github.com/99designs/gqlgen)?

Here are a few links relevant to the topic of thread safe random number generation in Go:
- https://qqq.ninja/blog/post/fast-threadsafe-randomness-in-go/
- https://github.com/golang/go/issues/21393
- https://github.com/golang/go/issues/3611

If you don't care that there is contention on a locked random number generator across go-routines, use this: https://pkg.go.dev/golang.org/x/exp/rand#example-LockedSource. I will update the README with this recommendation, since it seems most users of `ulid` don't want to define their own thread safe entropy source, nor manage a `sync.Pool`.

If you find there's too much contention in your program in production (via profiling with `pprof`), you can implement the `sync.Pool` solution.

**Comment by tsenart** [comment]
I opened https://github.com/oklog/ulid/pull/80 to improve the README.

**Comment by tsenart** [comment]
As a follow up, here's a proposal for a `ulid.Make()` function that has sane defaults and is safe for concurrent use: https://github.com/oklog/ulid/pull/81

**Comment by frederikhors** [comment]
The method [described here](https://qqq.ninja/blog/post/fast-threadsafe-randomness-in-go/#using-hashmaphash) is really fast! Can we use it with ulid?

**Comment by peterbourgon** [comment]
> The method [described here](https://qqq.ninja/blog/post/fast-threadsafe-randomness-in-go/#using-hashmaphash) is really fast! Can we use it with ulid?

Of course. Just create an io.Reader that uses this as a source of entropy.

**Comment by frederikhors** [comment]
> > The method [described here](https://qqq.ninja/blog/post/fast-threadsafe-randomness-in-go/#using-hashmaphash) is really fast! Can we use it with ulid?
> 
> Of course. Just create an io.Reader that uses this as a source of entropy.

Can you show me an example please?

**Comment by tsenart** [comment]
Merged #81
